File: exercise1-leg_kinematics/matlab/Ex1.m
%% Q1
% Write down the rotation matrices as functions of the angles
% alpha, beta, gamma using anonymous functions 
% https://www.mathworks.com/help/matlab/matlab_prog/anonymous-functions.html
% Hint: R_B1 = @(alpha) [1,0,0; 0, cos(alpha), -sin(alpha); ... ];

R_B1 = @(alpha) ... ;
R_12 = @(beta)  ... ;
R_23 = @(gamma) ... ;



%% Q2
% Write down the 3x1 relative position vectors for link lengths l_i=1
r_3F_in3 = ...;
r_23_in2 = ...;
r_12_in1 = ...;
r_B1_inB = ...;

% Write down the homogeneous transformations
% Hint: Can be created as compound matrices: [R_XY(gamma), r_XY_inX; 0 0 0 1];
H_23 = @(gamma) ...;
H_12 = @(beta) ...;
H_B1 = @(alpha) ...;

% Create the cumulative transformation matrix
% We will assume input of the configuration vector q = [alpha, beta,
% gamma]'
% Hint: H_B3 is a product of the matrices above
H_B3 = @(q) ...; 

% find the foot point position vector
% Hint: This H_cut function just cuts out the first three rows of an H
% matrix to help recover a 3*1 vector. 
% Then, r_XY_inX = @(q) H_cut(H_XZ(q))*[r_ZY_inZ; 1];
H_cut = @(H) H(1:3,:);
r_BF_inB = @(q) ...;


%% Q3

% Calculate the foot point Jacobian as a fn of configuration vector 
% q = [alpha; beta; gamma]'
J_BF_inB = @(q) [...];
    
% what generalized velocity dq do you have to apply in a configuration q = [0;60°;-120°]
% to lift the foot in vertical direction with v = [0;0;-1m/s];
dr = [0; 0; -1];
qval = pi/180*([0; 60; -120]);

dq = ...;

fprintf('Q3: Target velocity r_dot = [%0.1f; %0.1f; %0.1f] m/s\n', dr(1), dr(2), dr(3));
fprintf('in current configuration q = [%0.1f; %0.1f; %0.1f] deg\n', qval(1)*180/pi, qval(2)*180/pi, qval(3)*180/pi);
fprintf('Requires qdot = [%0.1f; %0.1f; %0.1f] deg/s\n', dq(1)*180/pi, dq(2)*180/pi, dq(3)*180/pi);
fprintf('\n\n');


%% Q4

% write an algorithm for the inverse kinematics problem to
% find the generalized coordinates q that gives the end effector position 
% rGoal = [0.2,0.5,-2]' and store it in qGoal
q0 = pi/180*([0;-30;60]);
rGoal = [0.2;0.5;-2];

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%% enter here your algorithm
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        
...
 



%% Q5

% Write an algorithm for the inverse differential kinematics problem to
% find the generalized velocities dq to follow a circle in the body xz plane
% around the start point rCenter with a radius of r=0.5 and a 
% frequeny of 1Hz. The start configuration is q =  pi/180*([0,-60,120])'
q0 = pi/180*([0,-60,120])';
dq0 = zeros(3,1);
rCenter = r_BF_inB(q0);
radius = 0.5;
f = 0.25;
rGoal = @(t) rCenter + radius*[sin(2*pi*f*t),0,cos(2*pi*f*t)]';
drGoal = @(t) 2*pi*f*radius*[cos(2*pi*f*t),0,-sin(2*pi*f*t)]';

% define here the time resolution
deltaT = 0.01;
timeArr = 0:deltaT:1/f;

% q, r, and rGoal are stored for every point in time in the following arrays
qArr = zeros(3,length(timeArr));
rArr = zeros(3,length(timeArr));
rGoalArr = zeros(3,length(timeArr));

q = q0;
dq = dq0;
for i=1:length(timeArr)
    t = timeArr(i);
    % data logging, don't change this!
    q = q+deltaT*dq;
    qArr(:,i) = q;
    rArr(:,i) = r_BF_inB(q);
    rGoalArr(:,i) = rGoal(t);
    
    % controller: 
    % step 1: create a simple p controller to determine the desired foot
    % point velocity
    v = ...
    % step 2: perform inverse differential kinematics to calculate the
    % gneralized velocities
    dq = ...
    
end

plotTrajectory(timeArr, qArr, rArr, rGoalArr, true);

File: exercise1-leg_kinematics/python/Ex1.py
import numpy as np
import matplotlib.pyplot as plt
import Ex1_extras


#############################
####### QUESTION 1 ##########
#############################
# write down the rotation matrices as functions of alpha, beta, gamma
def R_B1(alpha):
    return np.array(
        [
            [1, 0, 0],
            [0, np.cos(alpha), -np.sin(alpha)],
            [0, np.sin(alpha), np.cos(alpha)],
        ]
    )


def R_12(beta):
    return np.array(
        [[np.cos(beta), 0, np.sin(beta)], [0, 1, 0], [-np.sin(beta), 0, np.cos(beta)]]
    )


def R_23(gamma):
    return np.array(
        [
            [np.cos(gamma), 0, np.sin(gamma)],
            [0, 1, 0],
            [-np.sin(gamma), 0, np.cos(gamma)],
        ]
    )


#############################
####### QUESTION 2 ##########
#############################
# write down the 3x1 relative position vectors for link lengths l_i=1
r_3F_in3 = np.array([[0], [0], [-1]])
r_23_in2 = np.array([[0], [0], [-1]])
r_12_in1 = np.array([[0], [0], [-1]])
r_B1_inB = np.array([[0], [1], [0]])


def create_H(R, r):
    # This is a helper function to create the (4,4) homogeneous transform matrix
    # from a (3,3) rotation matrix R and a (3,1) position vector r
    assert R.shape == (3, 3), "Rotation matrix must be size (3,3)"
    assert r.shape == (3, 1), "Relative position vector must be size (3,1)"
    return np.block([[R, r], [0, 0, 0, 1]])


def H_B3(q):
    # Compose the full chain homogeneous transform between the body and 3 frames
    assert q.shape == (3, 1), "Configuration vector q must be size (3,1)"
    alpha, beta, gamma = q[0, 0], q[1, 0], q[2, 0]
    H_B1 = create_H(R_B1(alpha), r_B1_inB)
    H_12 = create_H(R_12(beta), r_12_in1)
    H_23 = create_H(R_23(gamma), r_23_in2)
    H_B3 = H_B1 @ H_12 @ H_23
    return H_B3


def r_BF_inB(q):
    # Calculate the foot point position vector in body frame for a given
    # configuration q
    assert q.shape == (3, 1), "Configuration vector q must be size (3,1)"
    r = H_B3(q) @ np.vstack((r_3F_in3, [1]))
    return r[:3]


#############################
####### QUESTION 3 ##########
#############################


# Determine the foot point Jacobian J_BF_inB = d(B_r_BF_inB)/dq
def J_BF_inB(q):
    assert q.shape == (3, 1), "Configuration vector q must be size (3,1)"
    alpha, beta, gamma = q[0, 0], q[1, 0], q[2, 0]
    J = np.array(
        [
            [0, -np.cos(beta + gamma) - np.cos(beta), -np.cos(beta + gamma)],
            [
                np.cos(alpha) * (np.cos(beta + gamma) + np.cos(beta) + 1),
                -np.sin(alpha) * (np.sin(beta + gamma) + np.sin(beta)),
                -np.sin(alpha) * np.sin(beta + gamma),
            ],
            [
                np.sin(alpha) * (np.cos(beta + gamma) + np.cos(beta) + 1),
                np.cos(alpha) * (np.sin(beta + gamma) + np.sin(beta)),
                np.cos(alpha) * np.sin(beta + gamma),
            ],
        ]
    )
    return J


# Check that J is correct for a random configuration using the numerical checker
q_rand = np.random.rand(3, 1) * 2 * np.pi
Ex1_extras.J_check(J_BF_inB, r_BF_inB, q_rand)

# What generalized velocity dq do you have to apply in a configuration q = [0;60°;-120°]
# to lift the foot in vertical direction with v = [0;0;-1] m/s
dr = np.array([[0], [0], [-1]])
qval = np.pi / 180 * np.array([[0], [60], [-120]])

dq = np.linalg.inv(J_BF_inB(qval)).dot(dr)
print("Ex 3: Generalised foot velocity dq = {0}.T deg/s".format(dq.T / np.pi * 180))

#############################
####### QUESTION 4 ##########
#############################
# Write an algorithm for the inverse kinematics problem to find the generalized
# coordinates q that gives the end effector position rGoal = [0.2,0.5,-2]'
q0 = np.pi / 180 * np.atleast_2d([0, -30, 60]).T
r_goal = np.atleast_2d([0.2, 0.5, -2]).T

for i in range(10):
    r = r_BF_inB(q0)
    q0 = q0 + np.linalg.pinv(J_BF_inB(q0)).dot(r_goal - r)

print(
    "Ex 4: Inverse kinematics for r = {0}.T: q = {1}.T deg".format(
        r_goal.T, q0.T / np.pi * 180
    )
)

#############################
####### QUESTION 5 ##########
#############################
# Write an algorithm for the inverse differential kinematics problem to
# find the generalized velocities dq to follow a circle in the body xz plane
# around the start point rCenter with a radius of r=0.5 and a
# frequency of 1Hz. The start configuration is q =  pi/180*([0,-60,120])'
q0 = np.pi / 180 * np.atleast_2d([0, -60, 120]).T
dq0 = np.zeros((3, 1))
rCenter = r_BF_inB(q0)
radius = 0.5
f = 0.25

rGoal = (
    lambda t: rCenter
    + radius
    * np.atleast_2d([np.sin(2 * np.pi * f * t), 0, np.cos(2 * np.pi * f * t)]).T
)
drGoal = (
    lambda t: 2
    * np.pi
    * f
    * radius
    * np.atleast_2d([np.cos(2 * np.pi * f * t), 0, -np.sin(2 * np.pi * f * t)]).T
)

# define here the time resolution and controller gain Kp
# Try varying the gain to see the effect!
deltaT = 0.01
timeArr = np.arange(0, 1 / f, deltaT)
Kp = 5.0

# q, r, and rGoal are stored for every point in time in the following arrays
qArr = np.zeros((3, len(timeArr)))
rArr = np.zeros((3, len(timeArr)))
rGoalArr = np.zeros((3, len(timeArr)))

q = q0
dq = dq0
for i in range(len(timeArr)):
    t = timeArr[i]

    q = q + deltaT * dq
    qArr[:, [i]] = q
    rArr[:, [i]] = r_BF_inB(q)
    rGoalArr[:, [i]] = rGoal(t)

    # controller:
    # step 1: create a simple p controller to determine the desired foot
    # point velocity
    v = drGoal(t) + Kp * (rGoalArr[:, [i]] - rArr[:, [i]])
    # step 2: perform inverse differential kinematics to calculate the
    # generalized velocities
    dq = dq + np.linalg.pinv(J_BF_inB(q)).dot(v)


with plt.style.context("ggplot"):
    tp = Ex1_extras.TrajectoryPlotter(timeArr, qArr, rArr, rGoalArr)
    an = tp.animation()
    plt.show()
